### PR TITLE
Recommit [RISCV] Update the interface of sifive vqmaccqoq (#74284)

### PR DESCRIPTION
**Pull request by 4vtomat**
The
spec(https://sifive.cdn.prismic.io/sifive/60d5a660-3af0-49a3-a904-d2bbb1a21517_int8-matmul-spec.pdf)
is updated.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/rvv/sf_vqmaccus_4x8x4.ll [hunk 1]
```diff
@@ -121,38 +121,36 @@ entry:
...
 }
 
-define <vscale x 16 x i32> @intrinsic_vqmaccus_4x8x4_ta_i32m8(<vscale x 16 x i32> %0, <vscale x 8 x i8> %1, <vscale x 64 x i8> %2, iXLen %3) nounwind {
+define <vscale x 16 x i32> @intrinsic_vqmaccus_4x8x4_ta_i32m8(<vscale x 16 x i32> %0, <vscale x 8 x i8> %1, <vscale x 32 x i8> %2, iXLen %3) nounwind {
 ; CHECK-LABEL: intrinsic_vqmaccus_4x8x4_ta_i32m8:
 ; CHECK:       # %bb.0: # %entry
-; CHECK-NEXT:    vl8r.v v24, (a0)
-; CHECK-NEXT:    vsetvli zero, a1, e8, m1, ta, ma
-; CHECK-NEXT:    sf.vqmaccus.4x8x4 v8, v16, v24
+; CHECK-NEXT:    vsetvli zero, a0, e8, m1, ta, ma
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/sf_vqmaccus_4x8x4.ll:146]: The vsetvli on every test case is for `e8, m1`. I think you're not setting the `VLMul` field in the pseudoinstruction in tblgen

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/sf_vqmaccus_4x8x4.ll:146]: The same issue exists for sf.vfwmacc.4x4x4

Comment (4vtomat) [reply llvm/test/CodeGen/RISCV/rvv/sf_vqmaccus_4x8x4.ll:146]: Updated in this [commit](https://github.com/llvm/llvm-project/pull/76006)!

File: llvm/test/CodeGen/RISCV/rvv/sf_vfwmacc_4x4x4.ll [hunk 2]
```diff
@@ -13,7 +13,7 @@ declare <vscale x 1 x float> @llvm.riscv.sf.vfwmacc.4x4x4.nxv1f32.nxv4bf16.nxv1b
 define <vscale x 1 x float> @intrinsic_vfwmacc_4x4x4_tu_f32mf2(<vscale x 1 x float> %0, <vscale x 4 x bfloat> %1, <vscale x 1 x bfloat> %2, iXLen %3) nounwind {
 ; CHECK-LABEL: intrinsic_vfwmacc_4x4x4_tu_f32mf2:
 ; CHECK:       # %bb.0: # %entry
-; CHECK-NEXT:    vsetvli zero, a0, e16, mf4, tu, ma
+; CHECK-NEXT:    vsetvli zero, a0, e16, mf2, tu, ma
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/sf_vfwmacc_4x4x4.ll:16]: mf4 was correct. It should match the <vscale x 1 x bfloat> operand

File: llvm/test/CodeGen/RISCV/rvv/sf_vqmacc_4x8x4.ll [hunk 3]
```diff
@@ -7,10 +7,10 @@
...
   <vscale x 8 x i8>,
-  <vscale x 8 x i8>,
+  <vscale x 4 x i8>,
   iXLen, iXLen);
 
-define <vscale x 2 x i32> @intrinsic_vqmacc_4x8x4_tu_i32m1(<vscale x 2 x i32> %0, <vscale x 8 x i8> %1, <vscale x 8 x i8> %2, iXLen %3) nounwind {
+define <vscale x 2 x i32> @intrinsic_vqmacc_4x8x4_tu_i32m1(<vscale x 2 x i32> %0, <vscale x 8 x i8> %1, <vscale x 4 x i8> %2, iXLen %3) nounwind {
 ; CHECK-LABEL: intrinsic_vqmacc_4x8x4_tu_i32m1:
 ; CHECK:       # %bb.0: # %entry
 ; CHECK-NEXT:    vsetvli zero, a0, e8, m1, tu, ma
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/sf_vqmacc_4x8x4.ll:16]: This should be mf2. since vs2 is `<vscale x 4 x i8>` (LMUL=1/2)

**Review by topperc — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td [hunk 4]
```diff
@@ -349,16 +349,24 @@ multiclass VPseudoSiFiveVMACC<string mx, VReg vd_type, VReg vs2_type,
...
+multiclass VPseudoSiFiveVQMACCDOD<string Constraint = ""> {
   foreach m = MxListVF8 in
     let VLMul = m.value in
     defm NAME : VPseudoSiFiveVMACC<m.MX, m.vrclass, m.vrclass, Constraint>;
 }
 
+multiclass VPseudoSiFiveVQMACCQOQ<string Constraint = ""> {
+  foreach i = [0, 1, 2, 3] in
+    let VLMul = MxListVF4[i].value in
+    defm NAME : VPseudoSiFiveVMACC<MxListVF8[i].MX, MxListVF8[i].vrclass,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:361]: The .MX passed to the first template argument should match VLMul. It's used to form the name of the PseudoInstruction. That name should be the same as the VLMul value.

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td [hunk 5]
```diff
@@ -349,16 +349,24 @@ multiclass VPseudoSiFiveVMACC<string mx, VReg vd_type, VReg vs2_type,
...
+}
+
 multiclass VPseudoSiFiveVFWMACC<string Constraint = ""> {
-  foreach m = MxListFW in
-    let VLMul = m.value in
-    defm NAME : VPseudoSiFiveVMACC<m.MX, m.wvrclass, m.vrclass, Constraint>;
+  foreach i = [0, 1, 2, 3, 4] in
+    let VLMul = MxListVF2[i].value in
+    defm NAME : VPseudoSiFiveVMACC<MxListVF4[i].MX, MxListFW[i].wvrclass,
+                                   MxListFW[i].vrclass, Constraint>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:369]: Is MxListVF2[i] and MxListFW[i] the same?

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td (hunk 4 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:361]: Got it~

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td (hunk 5 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:369]: Yes, it's the same lol~

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td [hunk 6]
```diff
@@ -553,29 +560,40 @@ class GetFTypeInfo<int Sew> {
 }
 
 multiclass VPatVMACC<string intrinsic, string instruction, string kind,
-                     list<VTypeInfoToWide> info_pairs, ValueType vec_m1> {
+                     list<VTypeInfoToWide> info_pairs, ValueType vec_m1,
+                     bit lmul_follows_vd = 0> {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:564]: lmul doesn't follow vd for any of these instructions. It always follows vs2, but one of the instructions has the same lmul for vs2 and vd.

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td [hunk 7]
```diff
@@ -349,14 +349,21 @@ multiclass VPseudoSiFiveVMACC<string mx, VReg vd_type, VReg vs2_type,
...
 
-multiclass VPseudoSiFiveVQMACC<string Constraint = ""> {
+multiclass VPseudoSiFiveVQMACCDOD<string Constraint = ""> {
   foreach m = MxListVF8 in
     let VLMul = m.value in
     defm NAME : VPseudoSiFiveVMACC<m.MX, m.vrclass, m.vrclass, Constraint>;
 }
 
+multiclass VPseudoSiFiveVQMACCQOQ<string Constraint = ""> {
+  foreach i = [0, 1, 2, 3] in
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:359]: `foreach i = 0-3 in`

File: llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td (hunk 6 above)
Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXSf.td:564]: Oh, that's right.

**Review by topperc — APPROVED** [review]
Lgtm

**Pull request merged**